### PR TITLE
SUGGESTION: Allow CNAME records on A/AAAA lookups

### PR DESCRIPTION
**Pull request by balboah**
This resolves my specific issue mentioned in #184. But I haven't considered what other records that should allow CNAME records and the risk involved with returning too many records or breaking other parts of trust-dns.

@bluejekyll

### CODE REVIEW

**Comment by balboah** [comment]
I noticed the cname ttl test failed. I'm not sure if this should really be there after these changes. The CNAME records have their own TTLs and I don't think that the A record TTL would need to be changed?

File: .rustfmt.toml [hunk 1]
```diff
@@ -1,3 +1,4 @@
+edition = "2018"
```
Comment (bluejekyll) [inline .rustfmt.toml]: did this make any difference?

File: crates/resolver/src/lookup_state.rs [hunk 2]
```diff
@@ -290,19 +290,27 @@ impl<C: DnsHandle + Send + 'static> CachingClient<C> {
                 .filter_map(|r| {
                     // because this resolved potentially recursively, we want the min TTL from the chain
                     let ttl = cname_ttl.min(r.ttl());
-
                     // TODO: disable name validation with ResolverOpts? glibc feature...
                     // restrict to the RData type requested
                     if query.query_class() == r.dns_class() {
                         // standard evaluation, it's an any type or it's the requested type and the search_name matches
-                        // - or -
+                        if (query.query_type().is_any() || query.query_type() == r.rr_type())
```
Comment (bluejekyll) [inline crates/resolver/src/lookup_state.rs]: This looks good.

I'm wondering if we should add an option to the always include, or not, and enable it by default only for the ResolverAuthority, but off for backward compatibility?

Adding it to `ResolverOpts` as perhaps a `preserve_intermediates`, which could just hang onto CNAMEs or potentially other future intermediate records? That might give a nice balance between the cost of storing them, vs. the desire to have them for specific scenarios?

File: .rustfmt.toml (hunk 1 above)
Comment (balboah) [reply .rustfmt.toml]: When I ran `rustfmt` to resolve the formatting check error in your tests, it complained about the `async fn's`. After specifying this edition the error went away

**Comment by balboah** [comment]
regarding the test failure, it was failing because the test iterates all records while assuming they all be the lower TTL. The test assumed we wouldn't return the additional CNAMEs which may have higher TTL

**Comment by balboah** [comment]
regarding `ResolverOpts` to toggle the feature flag for CNAME intermediates: It might be a good idea for preserving behaviour for the users who doesn't care about this forwarder issue. Would you mind adding that change? I'm still a bit lost on the whole setup, even though I managed to find this part of the code where CNAMEs may be returned :)

**Comment by balboah** [comment]
I tried to fix the integration test for *test_max_chained_lookup_depth* but it seems to still be failing, I'll leave it like this for today. Feel free to add changes in this branch regarding the ResovlerOpts

**Comment by bluejekyll** [comment]
Ok, I'm going to pull this down and finish it off.

**Comment by bluejekyll** [comment]
@balboah I've merge this with #1033

Thanks for the help!

**Pull request closed without merging**